File: rgctl-tests/ecommerce-go/internal/service/order.go
package service

import (
	"fmt"

	"github.com/google/uuid"
	"github.com/sshaaf/ecommerce-go/internal/models"
	"github.com/sshaaf/ecommerce-go/internal/pkg/timeutil"
	"github.com/sshaaf/ecommerce-go/internal/repository"
)

type OrderService struct {
	orders    *repository.OrderRepository
	cart      *repository.CartRepository
	products  *repository.ProductRepository
	inventory *repository.InventoryRepository
}

func NewOrderService(
	orders *repository.OrderRepository,
	cart *repository.CartRepository,
	products *repository.ProductRepository,
	inventory *repository.InventoryRepository,
) *OrderService {
	return &OrderService{
		orders:    orders,
		cart:      cart,
		products:  products,
		inventory: inventory,
	}
}

type OrderResponse struct {
	ID         string             `json:"id"`
	Status     string             `json:"status"`
	TotalCents int64              `json:"total_cents"`
	Items      []models.OrderItem `json:"items"`
}

func (s *OrderService) Checkout(userID string) (*OrderResponse, error) {
	items, err := s.cart.ListForUser(userID)
	if err != nil {
		return nil, err
	}
	if len(items) == 0 {
		return nil, NewBadRequest("cart is empty")
	}

	orderID := uuid.NewString()
	var total int64
	orderItems := make([]models.OrderItem, 0, len(items))

	for _, item := range items {
		product, err := s.products.FindByID(item.ProductID)
		if err != nil {
			return nil, MapRepoError(err)
		}
		if product.Stock < item.Quantity {
			return nil, NewBadRequest(fmt.Sprintf("insufficient stock for %s", product.Name))
		}
		total += product.PriceCents * item.Quantity
		orderItems = append(orderItems, models.OrderItem{
			ID:             uuid.NewString(),
			OrderID:        orderID,
			ProductID:      product.ID,
			Quantity:       item.Quantity,
			UnitPriceCents: product.PriceCents,
		})
	}

	order := &models.Order{
		ID:         orderID,
		UserID:     userID,
		Status:     "confirmed",
		TotalCents: total,
		CreatedAt:  timeutil.NowISO(),
	}
	if err := s.orders.Create(order); err != nil {
		return nil, err
	}
	for _, oi := range orderItems {
		if err := s.inventory.DecrementStock(oi.ProductID, oi.Quantity); err != nil {
			return nil, NewBadRequest(err.Error())
		}
		if err := s.orders.AddItem(&oi); err != nil {
			return nil, err
		}
	}
	if err := s.cart.Clear(userID); err != nil {
		return nil, err
	}
	return &OrderResponse{
		ID:         order.ID,
		Status:     order.Status,
		TotalCents: order.TotalCents,
		Items:      orderItems,
	}, nil
}

func (s *OrderService) List(userID string) ([]OrderResponse, error) {
	orders, err := s.orders.ListForUser(userID)
	if err != nil {
		return nil, err
	}
	out := make([]OrderResponse, 0, len(orders))
	for _, o := range orders {
		items, err := s.orders.ItemsForOrder(o.ID)
		if err != nil {
			return nil, err
		}
		out = append(out, OrderResponse{
			ID:         o.ID,
			Status:     o.Status,
			TotalCents: o.TotalCents,
			Items:      items,
		})
	}
	return out, nil
}

func (s *OrderService) Get(userID, id string) (*OrderResponse, error) {
	order, err := s.orders.FindByID(id)
	if err != nil {
		return nil, MapRepoError(err)
	}
	if order.UserID != userID {
		return nil, NewUnauthorized()
	}
	items, err := s.orders.ItemsForOrder(order.ID)
	if err != nil {
		return nil, err
	}
	return &OrderResponse{
		ID:         order.ID,
		Status:     order.Status,
		TotalCents: order.TotalCents,
		Items:      items,
	}, nil
}

File: rgctl-tests/ecommerce-rust/src/services/order.rs
use uuid::Uuid;
use crate::{
    dto::order::OrderResponse,
    error::{AppError, AppResult},
    models::order::{Order, OrderItem},
    repositories::{cart, inventory, order as repo, product},
    state::AppState,
    utils::time,
};

pub async fn checkout(state: &AppState, user_id: &str) -> AppResult<OrderResponse> {
    let items = cart::list_for_user(&state.pool, user_id).await?;
    if items.is_empty() { return Err(AppError::BadRequest("cart is empty".into())); }

    let order_id = Uuid::new_v4().to_string();
    let mut total = 0i64;
    let mut order_items = Vec::new();

    for item in &items {
        let p = product::find_by_id(&state.pool, &item.product_id).await?.ok_or(AppError::NotFound)?;
        if p.stock < item.quantity {
            return Err(AppError::BadRequest(format!("insufficient stock for {}", p.name)));
        }
        total += p.price_cents * item.quantity;
        order_items.push(OrderItem {
            id: Uuid::new_v4().to_string(),
            order_id: order_id.clone(),
            product_id: p.id.clone(),
            quantity: item.quantity,
            unit_price_cents: p.price_cents,
        });
    }

    let order = Order { id: order_id.clone(), user_id: user_id.into(), status: "confirmed".into(), total_cents: total, created_at: time::now_iso() };
    repo::create(&state.pool, &order).await?;
    for oi in &order_items {
        inventory::decrement_stock(&state.pool, &oi.product_id, oi.quantity).await?;
        repo::add_item(&state.pool, oi).await?;
    }
    cart::clear(&state.pool, user_id).await?;
    Ok(OrderResponse { id: order.id, status: order.status, total_cents: order.total_cents, items: order_items })
}

pub async fn list(state: &AppState, user_id: &str) -> AppResult<Vec<OrderResponse>> {
    let mut out = Vec::new();
    for o in repo::list_for_user(&state.pool, user_id).await? {
        let items = repo::items_for_order(&state.pool, &o.id).await?;
        out.push(OrderResponse { id: o.id, status: o.status, total_cents: o.total_cents, items });
    }
    Ok(out)
}

pub async fn get(state: &AppState, user_id: &str, id: &str) -> AppResult<OrderResponse> {
    let o = repo::find_by_id(&state.pool, id).await?.ok_or(AppError::NotFound)?;
    if o.user_id != user_id { return Err(AppError::Unauthorized); }
    let items = repo::items_for_order(&state.pool, &o.id).await?;
    Ok(OrderResponse { id: o.id, status: o.status, total_cents: o.total_cents, items })
}
